Add address management smart contract for managing a favorites list

ezekiiel opened the pull request:
Small contract for managing a sorted list of addresses. Might be used
for keeping a list of favorite contracts.

Addresses are added in the form:

```rust
struct AddressItem {
	pub addr: Addr,
	pub priority: u32,
}
```

One may then query for the contract list and addresses will be
returned in sorted order with the highest priority ones returned
first.

Placing many, countless addresses into this contract may cause it to
become unusable as it will run into out of gas issues. The contract
will only become locked at the point when the number of addresses
becomes so enormous that a single `log(N)` operation over a B-Tree set
excedes the gas limit.

JakeHartnell commented on contracts/address-manager/src/contract.rs:
```diff
@@ -0,0 +1,109 @@
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetAddresses {} => query_get_addresses(deps),
```
We may want pagination for this?

ezekiiel replied:
Good idea. Added optional start and end bounds to this query.

JakeHartnell commented on contracts/address-manager/src/error.rs:
```diff
@@ -0,0 +1,13 @@
+
+    #[error("Unauthorized")]
+    Unauthorized {},
+    // Add any other custom errors you like here.
+    // Look at https://docs.rs/thiserror/1.0.21/thiserror/ for details.
```
```suggestion
```

JakeHartnell commented on contracts/address-manager/rustfmt.toml:
```diff
@@ -0,0 +1,15 @@
+# stable
```
Do we need files like this if they exist in the root of this repo?

JakeHartnell commented on contracts/address-manager/LICENSE:
```diff
@@ -0,0 +1,202 @@
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
Update please with "DAO DAO" or yourself... @elsehow do we want a license with each contract?

ezekiiel replied:
Looks like we don't actually [need this](https://infra.apache.org/licensing-howto.html). Having the LICENCE file in the root seems like it is good enough

JakeHartnell commented on contracts/address-manager/.gitignore:
```diff
@@ -0,0 +1,15 @@
+# Build results
```
Do we need files like this if they exist in the root of this repo?

ezekiiel replied:
Good point. Probably not because this is a workspace and build output ends up in the root. Removing..

JakeHartnell commented on contracts/address-manager/.github/workflows/Basic.yml:
```diff
@@ -5,34 +5,33 @@ on: [push, pull_request]
 name: Basic
 
 jobs:
+
```
Feel like we only need one file like this in the repo root.

JakeHartnell commented on contracts/address-manager/.editorconfig:
```diff
@@ -0,0 +1,11 @@
+root = true
```
Do we need files like this if they exist in the root of this repo?

JakeHartnell commented on contracts/address-manager/Cargo.toml:
```diff
@@ -0,0 +1,43 @@
+[package]
+name = "address-manager"
```
Prefix with `cw-`?

JakeHartnell commented on contracts/address-manager/Cargo.toml:
```diff
@@ -0,0 +1,43 @@
+
+[dependencies]
+cosmwasm-std = { version = "1.0.0-beta4" }
+cosmwasm-storage = { version = "1.0.0-beta4" }
```
You can leave off the `4` here and it will use the latest version. For example: `"1.0.0-beta"`.

JakeHartnell commented on contracts/address-manager/src/contract.rs:
```diff
@@ -0,0 +1,117 @@
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetAddresses {start, end} => query_get_addresses(deps, start, end),
```
Maybe a query to check if a single address is in the list?

JakeHartnell replied:
`QueryMsg::CheckAddress { addr }`?

JakeHartnell reviewed (commented):
This is great! Just a few comments, and then I think this is ready. : )

ezekiiel commented:
Renamed to cw-address-manager and added a `CheckAddress` query.  Also tested using the DAO Up! DAO today. See most recent proposals [here](https://testnet.daodao.zone/dao/juno192c76yeywn7xe84ddw9jjtstecnztqhf4cg7cex07gnyxavchulqv7zp3d).

JakeHartnell reviewed (approved):
LGTM!

JakeHartnell commented:
Closing in favor of the `cw4-group` fork approach as we discussed @ezekiiel.
